stl/inc/__msvc_int128.hpp

```diff
@@ -0,0 +1,1155 @@
+        const bool _Three_word_den = __d >= 32;
+        __d &= 31;
+        uint32_t __u[5]{
+            static_cast<uint32_t>(_Num._Word[0] << __d),
+            static_cast<uint32_t>(_Num._Word[0] >> (32 - __d)),
+            static_cast<uint32_t>(_Num._Word[1] << __d),
+            static_cast<uint32_t>(_Num._Word[1] >> (32 - __d)),
+            0,
+        };
+        uint32_t __v[4] = {
+            static_cast<uint32_t>(_Den._Word[0] << __d),
+            static_cast<uint32_t>(_Den._Word[0] >> (32 - __d)),
+            static_cast<uint32_t>(_Den._Word[1] << __d),
+            static_cast<uint32_t>(_Den._Word[1] >> (32 - __d)),
+        };
+        if (__d != 0) {
+            __u[2] |= _Num._Word[0] >> (64 - __d);
+            __v[2] |= _Den._Word[0] >> (64 - __d);
+        }
+
+        uint32_t __q[2];
+        if (_Three_word_den) {
+            // 4-digit by 3-digit base 2^32 division
+            _Knuth_4_3_1_D(__u, reinterpret_cast<uint(&)[3]>(__v), __q);
+        } else {
+            // 4-digit by 4-digit base 2^32 division
+            _Knuth_4_3_1_D(__u, __v, reinterpret_cast<uint32_t(&)[1]>(__q));
+            __q[1] = 0;
+        }
+
+        return _Uint128{static_cast<uint64_t>(__q[1]) << 32 | __q[0]};
+#endif // defined(_M_X64) && !defined(_M_ARM64EC)
+    }
+
+    constexpr _Uint128& operator/=(const uint32_t _That) noexcept {
+        *this = *this / _That;
+        return *this;
+    }
+    constexpr _Uint128& operator/=(const uint64_t _That) noexcept {
+        *this = *this / _That;
+        return *this;
+    }
+    constexpr _Uint128& operator/=(const _Uint128& _That) noexcept {
+        *this = *this / _That;
+        return *this;
+    }
+    template <integral _Ty>
+    friend constexpr _Ty& operator/=(_Ty& _Left, const _Uint128& _Right) noexcept {
+        _Left = static_cast<_Ty>(_Uint128{_Left} / _Right);
+        return _Left;
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator%(const _Uint128& _Num, const uint32_t _Den) noexcept {
+        uint64_t _Rem = _Num._Word[1];
+        _Rem          = (_Rem % _Den) << 32 | (_Num._Word[0] >> 32);
+        _Rem          = (_Rem % _Den) << 32 | static_cast<uint32_t>(_Num._Word[0]);
+        return _Uint128{_Rem % _Den};
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator%(const _Uint128& _Num, const uint64_t _Den) noexcept {
+        uint64_t _Rem;
+        (void) _UDiv128(_Num._Word[1] % _Den, _Num._Word[0], _Den, _Rem);
+        return _Uint128{_Rem};
+    }
+
+    _NODISCARD friend constexpr _Uint128 operator%(_Uint128 _Num, _Uint128 _Den) noexcept {
+        // establish _Den < _Num and _Num._Word[1] > 0
+        if (_Den._Word[1] > _Num._Word[1]) {
+            return _Num;
+        }
+        if (_Den._Word[1] == _Num._Word[1]) {
+            if (_Den._Word[0] < _Num._Word[0]) {
+                return _Uint128{_Num._Word[0] - _Den._Word[0]};
+            }
+
+            return _Num;
+        }

```

Comment:
There is a bug here - when `_Num == _Den`, this returns `_Num` instead of `_Uint128{}`. Indicates missing test coverage for `X % X`.